File: src/main/java/com/raychenon/leetcode/RotatedDigits.java
package com.raychenon.leetcode;

/**
 * User: raychenon
 * Date: 2019-11-24
 * https://leetcode.com/problems/rotated-digits/
 */
public class RotatedDigits {

    public int rotatedDigits(int N) {
        int count = 0;
        for (int i = 1; i <= N; ++i) {
            if (isValid(i)) ++count;
        }
        return count;
    }

    private boolean isValid(int n) {
        StringBuilder str = new StringBuilder();

        int rem = n;  // remainder
        while (rem != 0) {
            int modulo = rem % 10;
            // invalid digits: 3,4,7
            if (modulo == 3 || modulo == 4 || modulo == 7) {
                return false;
            }
            modulo = reversedDigit(modulo);
            str.append(modulo);
            rem = rem / 10;
        }

        str.reverse();

        return !String.valueOf(n).equals(str.toString());
    }

    private int reversedDigit(int n) {
        switch (n) {
            case 2:
                return 5;
            case 5:
                return 2;
            case 6:
                return 9;
            case 9:
                return 6;
            default:
                return n;
        }
    }

}

File: src/main/java/com/raychenon/leetcode/RotatedDigits.kt
package com.raychenon.leetcode

/**
 * User: raychenon
 * Date: 2019-11-24
 * https://leetcode.com/problems/rotated-digits
 */
object RotateDigits {

    fun rotatedDigits(N: Int): Int {
        var count = 0
        for (i in 1..N) {
            if (isValid(i)) ++count
        }
        return count
    }


    // 0->0
    // 1->1
    // 8->8

    // 2<->5
    // 6<->9

    // invalid  3,4,7
    fun isValid(n: Int): Boolean {
        var nb = n
        var rotated = 0
        var str = StringBuffer()
        while (nb != 0) {
            var modulo = nb % 10
            // invalid
            if (modulo == 3 || modulo == 4 || modulo == 7) {
                // println("n = $n")
                return false
            }
            modulo = when (modulo) {
                2 -> 5
                5 -> 2
                6 -> 9
                9 -> 6
                else -> modulo
            }
            rotated = rotated * 10 + modulo
            str.append(modulo)
            nb = nb / 10
        }

        str.reverse()
        val res = !Integer.toString(n).equals(str.toString())
        //println("n = $n , rotated = $rotated , ${str.toString()} , ${res} ")
        return res
    }
}